node-support/src/metadata.js

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2019 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+function Metadata(entries) {
```

Comment:
@jroper Should this perhaps normalize entries to avoid semantically equivalent keys but added with different casing?

Reply:
The semantics of keys are dependent on the transport. In HTTP, they are case insensitive. In other transports, they might not be.

My aim here has been to ensure that the user experience is smooth for as wide a range of use cases as possible. One of the most common places to use metadata is HTTP, and HTTP header names are case insensitive. For transports where header names are case sensitive, I think it would be very, very unusual if those transports ever defined multiple header names that differed only by case. So, given this, I've made all lookups case insensitive, this satisfies at least 99% of use cases, it gives a good experience in HTTP, and for case sensitive transports, it shouldn't matter whether the lookups are case insensitive if keys almost never differ by case. Of course, in the once in the history of the universe chance that there will be keys that differ only by case, you can still iterate through the entries to differentiate between them, and you get the key in its original case.

For setting metadata - I preserve the case passed. Plus, multiple values for a single case are supported here, so this always adds the value in addition to old values, there's no replace operation. If you want to replace a key, you'll have to delete followed by add. Maintaining case when setting ensures that things work both in the case sensitive and case insensitive scenarios. It's also how most HTTP/1.1 clients and servers generally work, even though http header names are case insensitive, they always preserve the case passed to them by users.

Reply:
I think key is to make sure that it's difficult to mess things up, and if things get messed up, it's easy to diagnose and remedy. Having effectively a "bag"-semantics I think is rather error prone, wouldn't you agree?

Reply:
It's technically not bag semantics because insertion order is maintained.

But yes, I agree it is error prone - but if you don't like that, the person to take it up with is Tim Berners-Lee. The error proneness doesn't come from what we're providing, it is inherited from what we are supporting. We're supporting HTTP as a transport protocol. HTTP headers have bag-like semantics, which is error prone, supporting HTTP headers means inheriting those error prone semantics. If we tried to fix it, the result would be a metadata structure where certain valid HTTP messages couldn't be expressed. And given that metadata is an escape hatch into the transport protocol, it wouldn't make sense to me if the escape hatch only got you half way out.

Reply:
Talking to Tim, brb.